isa/kernel.org/instruction-set.rst

```diff
@@ -256,59 +379,64 @@ by other eBPF programs or means outside of this specification.
-  * 8-bit and 16-bit wide atomic operations are not supported.
+  * ``BPF_ATOMIC | BPF_W | BPF_STX`` (0xc3) for 32-bit operations
+  * ``BPF_ATOMIC | BPF_DW | BPF_STX`` (0xdb) for 64-bit operations
+
+Note that 8-bit (``BPF_B``) and 16-bit (``BPF_H``) wide atomic operations are not supported,
+nor is ``BPF_ATOMIC | <size> | BPF_ST``.
 
-The imm field is used to encode the actual atomic operation.
+The 'imm' field is used to encode the actual atomic operation.
 Simple atomic operation use a subset of the values defined to encode
-arithmetic operations in the imm field to encode the atomic operation:
+arithmetic operations in the 'imm' field to encode the atomic operation:
 
-  ========  =====  ===========
-  imm       value  description
-  ========  =====  ===========
-  BPF_ADD   0x00   atomic add
-  BPF_OR    0x40   atomic or
-  BPF_AND   0x50   atomic and
-  BPF_XOR   0xa0   atomic xor
-  ========  =====  ===========
+  ========  =====  ===========  =======
+  imm       value  description  version
+  ========  =====  ===========  =======
+  BPF_ADD   0x00   atomic add   1
+  BPF_OR    0x40   atomic or    3
+  BPF_AND   0x50   atomic and   3
+  BPF_XOR   0xa0   atomic xor   3
+  ========  =====  ===========  =======
 
+where 'version' indicates the first ISA version in which the value was supported.
 
-``BPF_ATOMIC | BPF_W  | BPF_STX`` with imm = BPF_ADD means::
+``BPF_ATOMIC | BPF_W  | BPF_STX`` with 'imm' = BPF_ADD means::
 
-  *(u32 *)(dst_reg + off16) += src_reg
+  *(uint32_t *)(dst + offset) += src
 
-``BPF_ATOMIC | BPF_DW | BPF_STX`` with imm = BPF ADD means::
+``BPF_ATOMIC | BPF_DW | BPF_STX`` with 'imm' = BPF ADD means::
 
-  *(u64 *)(dst_reg + off16) += src_reg
+  *(uint64_t *)(dst + offset) += src
 
-``BPF_XADD`` is a deprecated name for ``BPF_ATOMIC | BPF_ADD``.
+``BPF_XADD`` appeared in version 1, but is now considered to be a deprecated alias
+for ``BPF_ATOMIC | BPF_ADD``.
 
-In addition to the simple atomic operations, there also is a modifier and
+In addition to the simple atomic operations above, there also is a modifier and
 two complex atomic operations:
 
-  ===========  ================  ===========================
-  imm          value             description
-  ===========  ================  ===========================
-  BPF_FETCH    0x01              modifier: return old value
-  BPF_XCHG     0xe0 | BPF_FETCH  atomic exchange
-  BPF_CMPXCHG  0xf0 | BPF_FETCH  atomic compare and exchange
-  ===========  ================  ===========================
+  ===========  ================  ===========================  =======
+  imm          value             description                  version
+  ===========  ================  ===========================  =======
+  BPF_FETCH    0x01              modifier: return old value   3
+  BPF_XCHG     0xe0 | BPF_FETCH  atomic exchange              3
+  BPF_CMPXCHG  0xf0 | BPF_FETCH  atomic compare and exchange  3
+  ===========  ================  ===========================  =======
```

Comment:
Linux also has bitwise fetch-and-{and,or,xor} (values `BPF_AND | BPF_FETCH`, `BPF_OR | BPF_FETCH`, `BPF_XOR | BPF_FETCH` respectively). I have a quick reference [here](https://github.com/iovisor/bpf-docs/pull/26/commits/c41e3bb37b0f7d0bedd654bbd9e60246c62b669c).

Reply:
Line 425 already says `BPF_FETCH` can be used with any of the simple operations above.
And they appear in the appendix in lines 617-619, etc.

Reply:
OK my bad